sql/handler.cc

```diff
@@ -6356,12 +6356,43 @@ static bool create_sys_trx_field_if_missing(THD *thd, const char *field_name,
   return false;
 }
 
-bool System_versioning_info::add_implicit_fields(THD *thd,
+bool System_versioning_info::add_versioning_info(THD *thd,
                                                  Alter_info *alter_info)
 {
-  if (!declared_system_versioning)
+  // If user defined something related to versioning treat this table as
+  // versioned.
+  versioned= has_versioned_fields || has_unversioned_fields ||
+             declared_system_versioning || period_for_system_time.start ||
+             period_for_system_time.end || generated_as_row.start ||
+             generated_as_row.end;
+
+  if (!versioned)
```

Comment:
If you mean, that this will be in `add_versioning_info()`, then yes, it is acceptable. Now this `if (!versioned)` check looks abundant.